### DISCUSSION TITLE
How to download third party images from Front End

### DISCUSSION THREAD

**Discussion by NeeshSamsi** [vercel/next.js#41050] [Help] [answered]
I'm using Prismic as my CMS to serve images. I added the prismic images domain and displaying images works fine. I have a gallery section where I want users to be able to download said images. The problem is that the `<a href="" download>` download attribute does that work with third party images. I assume that I need to query an api endpoint that acts as a proxy to fetch the image and download it. I tried messing around with various blog solutions like:
https://www.grooovinger.com/blog/image-proxy-with-next-js/
https://stackoverflow.com/questions/68490546/how-to-download-a-file-on-next-js-using-an-api-route
https://stackoverflow.com/questions/70161372/how-to-download-an-image-and-then-serve-next-js-api-routes

I am also recently getting familiar with working with 'fs'
I assume I need to fetch and create the image file like so:
```
const file = fs.createWriteStream(`image.jpg`)
const request = https.get(imageUrl, function(response) {
  response.pipe(file).on('close', function () {
    console.log('File written!')
  });
res.setHeader("Content-Disposition", "attachment; filename=" + fileName);
```
I learned that I need the Header: 'Content-Disposition', 'attachment; filename=filename' to initiate a download but I am lost on:
1. Whether this will work?
2. How I can send the file const in the response to be downloaded

**Comment by icyJoseph** [comment, marked as answer]
Hi,

First of all, code to do this doesn't really need to create a stream like that. Not in the context we are working with anyway.

```ts
import type { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const picRes = await fetch("https://picsum.photos/200/300");

  console.log("content-type:", picRes.headers.get("content-type"));

  // maybe you can use this too see if the image's too large
  // to send downstream
  console.log("content-length:", picRes.headers.get("content-length")); 
  const imageBlob = await picRes.blob();

  const chunks: any[] = [];

  for (const chunk of imageBlob.stream().read()) {
    chunks.push(chunk);
  }

  console.log(chunks.length);
  res.setHeader(
    "content-type",
    picRes.headers.get("content-type") || "image/*"
  );

  res.setHeader(
    "content-length",
    picRes.headers.get("content-length") || chunks.length
  );

  res.setHeader("Content-Disposition", 'attachment; filename="tomato.jpeg"');

  res.write(Uint8Array.from(chunks));

  return res.status(200).end();
}
```

Second, there's a natural limitation of API routes, which is the payload they can allow through:

- https://nextjs.org/docs/messages/api-routes-response-size-limit
- https://vercel.com/docs/concepts/limits/overview#serverless-function-payload-size-limit
- https://nextjs.org/docs/api-routes/request-helpers#custom-config

So you ought to really consider, beyond the technical details, whether or not this is a good solution.

**Reply by NeeshSamsi** [reply]
Thank you SOOO much. I struggled for a while and posted this question on quite a few forums and discord servers. This is the answer that nailed it. Thanks again.

I checked out the Next.js API response size limitations. From what I learned, there is a 4mb limit by default. If I understood it correctly, the images I am downloading needs to be less than 4mb, which should be fine. Even if at some point, I need to download larger files, it seems I can change or remove the limit at the cost of response time.

**Reply by nagendragandla001** [reply]
Small Update on top of the code.
imageBlob.stream().read() will get an error as it's not part of ReadableStram<>
It should be
await imageBlob.stream().getReader().read();
